Add vmgroup tests

govint opened the pull request:
Added tests for vmgroups for #1262. Some tests are overlapped and hence not included. Most tests are done and will complete the remaining. Pls. review for structure and format.

Ideally, we should be going with a template so all *_test files look the same in terms of structure. I've refactored some code in places and will more common code into a separate e2emain_test.go shortly.

shuklanirdesh82 commented on tests/utils/dockercli/volumelifecycle.go:
```diff
@@ -26,17 +26,17 @@ import (
 func AttachVolume(ip, volName, containerName string) ([]byte, error) {
 	log.Printf("Attaching volume [%s] on VM [%s]\n", volName, ip)
-	return ssh.InvokeCommand(ip, dockercli.RunContainer+"-d -v "+volName+
+	return ssh.InvokeCommand(ip, dockercli.RunContainer+"-d -rm -v "+volName+
```
isn't this `--rm` instead of `-rm`?

shuklanirdesh82 commented on tests/e2e/vmgroup_volumeops_test.go:
```diff
@@ -0,0 +1,300 @@
+
+// TestVolumeCreateOnDefaultVg - Verify that volumes can be created on the
+// default volume group with default permissions, then attached and deleted
+func TestVolumeCreateAndAttachOnDefaultVg(t *testing.T) {
```
To keep it consistent (reference of any existing test), high level steps should be added per test method.

govint replied:
Done.

shuklanirdesh82 commented on tests/utils/ssh/ssh.go:
```diff
@@ -47,5 +45,16 @@ func InvokeCommandLocally(cmd string) string {
+
+//ExecCmd method takes command and host and calls InvokeCommand
+//and then returns the output after converting to string
+func ExecCmd(hostName string, cmd string) (string, error) {
```
Thanks for catching! .. ssh.go is the perfect home.

ashahi1 commented on tests/e2e/vmgroup_volumeops_test.go:
```diff
@@ -0,0 +1,300 @@
+	if err != nil {
+		t.Errorf("Error while creating volume group - [%s] err - %s", cmd, err.Error())
+	}
+
```
It would be better if we could also verify the new vmgroup created using admincli vmgroup ls.

shaominchen commented on tests/constants/admincli/cmd.go:
```diff
@@ -25,4 +25,28 @@ const (
 	ListVolumes = vmdkopsAdminVolume + "ls "
+
+	// vmdkopsCLIVolumegroup Volume group CLI commands
+	vmdkopsCLIVolumegroup = vmdkopsAdmin + "vmgroup"
```
1. Volumegroup should be VmGroup. Same for other constants below.
2. Let's have consistent prefix, either "vmdkopsAdmin" or "vmdkopsCLI", or simply without prefix since there's no naming conflicts in this const type.

govint replied:
Fixed. Volume group is wrong and have replaced as VmGroup

shaominchen commented on tests/e2e/vmgroup_volumeops_test.go:
```diff
@@ -0,0 +1,300 @@
+// After installing vmdk volume plugin/driver, volume creation should not be
+// failed very first time.
+
+// This test is going to cover the issue reported at #656
```
I'd suggest to give some brief introduction about the test suite, and add the issue number just as a reference.

govint replied:
Done.

shaominchen commented on tests/utils/dockercli/volumelifecycle.go:
```diff
@@ -26,17 +26,17 @@ import (
 
 // CreateVolume is going to create vsphere docker volume with given name.
-func CreateVolume(ip, name string) ([]byte, error) {
+func CreateVolume(ip, name, opts string) ([]byte, error) {
```
Quite some test cases simply need to create a volume with default configuration. So I'd suggest to change this to a variadic function just for convenience:

   func CreateVolume(ip, name string, opts ...string) ([]byte, error)

This way we don't need to modify existing test cases.

govint replied:
Added a CreateVolumeWithOpts(), a variadic function will need more work in CreateVolume() to parse the args and call the docker command

shaominchen commented on tests/e2e/vmgroup_volumeops_test.go:
```diff
@@ -0,0 +1,300 @@
+
+import (
+	"os"
+	"testing"
```
Why don't we use gocheck in all e2e tests so that it's easier to maintain the code and also to get a consistent test output? @shuklanirdesh82

shuklanirdesh82 replied:
Yeah correct. @govint, as mentioned please take an example of `basic_test` or `restart_test` (currently checked in and found at /tests/e2e package) use the gocheck library for new tests.

As per one of comment (https://github.com/vmware/docker-volume-vsphere/issues/472#issuecomment-226411645) test framework has been integrated to write our automation scripts.

govint commented on vmdk_plugin/Makefile:
```diff
@@ -350,7 +350,7 @@ test-all:  test e2e-test testremote
 e2e-test:
 	$(log_target)
-	$(SCRIPTS)/build.sh e2e-dkrVolDriver-test "$(ESX)" "$(VM1)" "$(VM2)"
+	$(SCRIPTS)/build.sh e2e-dkrVolDriver-test "$(ESX)" "$(VM1)" "$(VM2)" "$(DS1)" "$(DS2)" "$(CNAME)" "$(T1)"
```
Sure, can I get a reference to use this?

shuklanirdesh82 replied:
sharing some command for your reference. 

`govc vm.info -json <VM_NAME> | jq -r .VirtualMachines[].Config.DatastoreUrl[].Name`: grabbing VM's datastore value.

`govc datastore.info`: returns all available datastores in the inventory

sample output: VM on my local testbed is "photon1" created on "TestDatastore"
```
nshukla-mbp:~ nshukla$ govc vm.info -json photon1 | jq -r .VirtualMachines[].Config.DatastoreUrl[].Name
TestDatastore
nshukla-mbp:~ nshukla$ govc datastore.info
Name:        TestDatastore
  Path:      /ha-datacenter/datastore/TestDatastore
  Type:      VMFS
  URL:       /vmfs/volumes/4fc608aa-828f03c0-d4a2-d4ae52794a7a
  Capacity:  460.8 GB
  Free:      415.1 GB
```

We have just started adding some handy utils at https://github.com/vmware/docker-volume-vsphere/blob/master/tests/utils/govc/cmd.go

Let me know if there are any issues. I can supply a quick util if you want as a separate PR. I will be looking forward to your response.

govint replied:
I think the govc commands should be part of a script that runs these commands, parses the results and populates the env before running the automated tests. That way the user has the flexibility to change the script and run with different values or set their env with their choice of data stores and other values and run the test. It's easier to parse the govc values in script than in go code. What do you'll say.

shuklanirdesh82 replied:
Yeah we can do that. Aren't we going to add an extra overhead for any code contributor? 

IMO, the same we can achieve through testMain() that you have introduced before running any tests to keep it consistent by using go program only.  

Other approach I am thinking of maintaining consistency, it is better to have "on demand" queries instead of having a scripts which prepares the environment before hand, this approach is simpler while dealing with CI too. Only specific tests/testgroups invokes govc queries.

Going forward, we will be breaking down tests into separate areas (e.g. swarm, vmgroup, basic_test, plugin/refcount tests etc); with the minimal ask from user test would be able to run instead of preparing all possible environment variables.

shaominchen commented on tests/constants/admincli/cmd.go:
```diff
@@ -25,4 +25,28 @@ const (
 	// ListVolumes referring to vmdkops_admin volume ls
 	ListVolumes = vmdkopsAdminVolume + "ls "
+
+	// vmdkopsAdminVmGroup Volume group CLI commands
```
Volume group => VmGroup

govint replied:
Thanks, no idea why I even thought "volume groups" in the first place.

shaominchen commented on tests/e2e/vmgroup_volumeops_test.go:
```diff
@@ -0,0 +1,271 @@
+	if err != nil {
+		t.Errorf("Error while deleting volume - %s in the default , err - %s", volName1, err.Error())
+	}
+	t.Logf("Passed - create privilege on default ")
```
In existing tests, we are adding 2 tracing logs for each test:

log.Printf("START: <test-group>.<test-case>")
log.Printf("END: <test-group>.<test-case>")

Can we follow the same so that we have a consistent output?

govint replied:
Agree and will add.

shaominchen commented on tests/e2e/e2e_main.go:
```diff
@@ -0,0 +1,61 @@
+var DockerHosts = []string{os.Getenv("VM2"), os.Getenv("VM1")}
+var Host, Vmgroup, Ds1, Ds2, Cname string
+
+func TestMain(m *testing.M) {
```
Is this TestMain introduced for vmgroup_volumeops_test.go only? If the answer is yes, then shouldn't we move these code into vmgroup_volumeops_test.go, or maybe just give this file a better (more specific) name? The name "e2e_main" is too generic which should be used for the whole e2e tests only.

govint replied:
No, since all these tests are part of e2e package, this function Test
main() is common to all the tests. All globals are also common to all the
test code. Hence keeping env used in multiple tests in one location and
init of those in Testmain().

govint replied:
Removed this and using the infra for gocheck instead.

govint commented:
I think the govc commands should be pay of a script that runs these
commands, parses the results and populates the env before running the
automated tests. That way the user has the flexibility to change the script
and run with different values or set their env with their choice of data
stores and other values and run the test. It's easier to parse the govc
values in script than in go code. What do you'll say.

govint commented:
No, since all these tests are part of e2e package, this function Test
main() is common to all the tests. All globals are also common to all the
test code. Hence keeping env used in multiple tests in one location and
init of those in Testmain().

govint commented:
Addressed comments and pushed change - uses gocheck entirely for all tests. Govc can be done once this version is in and govc use is preferrably centralized into a single place in the test code so all tests can use values from govc cli.

shaominchen commented on tests/e2e/vmgroups_test.go:
```diff
@@ -0,0 +1,307 @@
+	}
+
+	if vg.datastore1 == "" || vg.datastore2 == "" {
+		os.Exit(1)
```
os.Exit(1) will exit the whole e2e test suite, right? We shouldn't use this as it will break other test groups as well. Each test group should be run independently without affecting other test groups.

govint replied:
Fixed

shaominchen commented on tests/e2e/vmgroups_test.go:
```diff
@@ -0,0 +1,307 @@
+)
+
+const (
+	volName1                 = "_test_vol_1"
```
In other test groups we are using inputparams.GetVolumeNameWithTimeStamp(test_group_name) to get the volume names. The reason we are using test_group_name here is to make debugging easier - when some tests fail and some volumes are being left over in the system, it will be easier to tell which test group did not clean up volumes properly, simply by reading the name.

govint replied:
Done.

shaominchen commented on tests/e2e/vmgroups_test.go:
```diff
@@ -0,0 +1,307 @@
+	vmgroup string
+	datastore1 string 
+	datastore2 string
+	cname string
```
What's cname - canonical name?

govint replied:
Nope, container name and changed it as such.

govint replied:
Container name - fixed.

shaominchen commented on tests/e2e/vmgroups_test.go:
```diff
@@ -0,0 +1,307 @@
+	volName1                 = "_test_vol_1"
+	volName2                 = "_test_vol_2"
+	defaultVg                = "_DEFAULT"
+	defcname                 = "def_test_ctr"
```
Nit: what does "ctr" mean here?

govint replied:
Container  - renamed and fixed.

shaominchen commented on tests/e2e/vmgroups_test.go:
```diff
@@ -0,0 +1,307 @@
+func (vg *VmGroupTest) SetUpSuite(c *C) {
+	vg.dockerHosts = []string{os.Getenv("VM1"), os.Getenv("VM2"), os.Getenv("VM3")}
+	vg.host = os.Getenv("ESX")
+	vg.vmgroup = os.Getenv("VMGROUP")
```
These variables are not available in the current CI environment. Don't we need to add them to the build scripts - similar to what we have done in PR #1292?

shuklanirdesh82 replied:
is there a reason for passing/setting as environment variable?

govint replied:
Other tests are using VM? env. vars, do we plan to change those as well?

shaominchen commented on tests/e2e/vmgroups_test.go:
```diff
@@ -0,0 +1,307 @@
+
+	// 2. Verify the volume is created on the default vm group
+	cmd := admincli.ListVolumes + " | grep " + defaultVg + " | grep " + name
+	val, err := ssh.ExecCmd(vg.host, cmd)
```
We should stop using Exec.Cmd - please see issue #1294 for the reason.

govint replied:
Fixed.

govint replied:
Fixed.

shuklanirdesh82 commented on tests/utils/ssh/ssh.go:
```diff
@@ -45,3 +45,13 @@ func InvokeCommandLocally(cmd string) string {
+
+//ExecCmd method takes command and host and calls InvokeCommand
+//and then returns the output after converting to string
+func ExecCmd(hostName string, cmd string) (string, error) {
```
with #1294 there is an ask for getting rid of this util. can you please take care as part of this change?

govint replied:
Fixed.

govint replied:
Fixed.

shuklanirdesh82 commented on tests/e2e/vmgroups_test.go:
```diff
@@ -0,0 +1,307 @@
+const (
+	volName1                 = "_test_vol_1"
+	volName2                 = "_test_vol_2"
+	defaultVg                = "_DEFAULT"
```
with https://github.com/vmware/docker-volume-vsphere/pull/1283/files#diff-a3a1723e96c4aa2a131a49f75ebd95e0R45 .. we will be getting as admincli constant, please avoid using local test constants.

govint replied:
I'll retain this for now till #1283 is merged.

shuklanirdesh82 commented on tests/e2e/vmgroups_test.go:
```diff
@@ -0,0 +1,307 @@
+	vg.dockerHosts = []string{os.Getenv("VM1"), os.Getenv("VM2"), os.Getenv("VM3")}
+	vg.host = os.Getenv("ESX")
+	vg.vmgroup = os.Getenv("VMGROUP")
+	vg.datastore1 = os.Getenv("DS1")
```
I would suggest letting code/util to retrieve available datastore (https://github.com/vmware/docker-volume-vsphere/pull/1286#discussion_r118540539) 

Rather than following stricter approach, we can let testcase author to decide which datastore to use from the available in the inventory.

shuklanirdesh82 commented:
@govint 

a lot of improved changes are there as part of this review, can you please run tests locally and update the PR with "Testing done" information? It will be easier to avoid obvious issue and review comments.

Thanks!

govint commented:
Yes, test code is evolving as the test are getting written. Maybe we could
define the common components and create a template that the tests could
use. There are still tests that are using the testing module vs. gocheck.
Once thats done the churn at individual test code level would be less.

@govint <https://github.com/govint>

a lot of improved changes are there as part of this review, can you please
run tests locally and update the PR with "Testing done" information? It
will be easier to avoid obvious issue and review comments.

Thanks!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/vmware/docker-volume-vsphere/pull/1286#issuecomment-304079286>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/APHseK3EIN9PhWEC2mQnTjf4mhFHRG-Wks5r9cHVgaJpZM4Njxd1>
.

shuklanirdesh82 commented:
> https://github.com/vmware/docker-volume-vsphere/pull/1286#issuecomment-304083642

@govint There are two testcase using go testing module. with #1296 one of the test will be removed and I have requested @ashahi1 to prioritize #1096 to use gocheck in 0.15 itself.

/CC @ashahi1

govint commented:
Test results:

1. The tests use the ESX, VM1, VM2 env vars.
1. The below env. vars are added and may be replaced later with govc calls.
1. VM1NAME - name of the VM whose IP is in VM1
1. VM2NAME - name of the VM whose IP is in VM2
1. DS1 - the datastore (basename) to use in addition to _VM_DS, as default DS for non-default vmgroup in multivmgroup tests.

Tried govc to get the vm name keeps failing with "connection reset by peer". Not reliable for an automated test.

make e2e-test ESX=10.110.48.56 VM1=10.110.48.54 VM2=10.110.48.50 VM1NAME=2 VM2NAME=1 DS1=datastore1
make --directory=vmdk_plugin e2e-test
make[1]: Entering directory `/home/administrator/vpl-test/docker-volume-vsphere/vmdk_plugin'
Makefile:152: warning: overriding commands for target `clean'
../plugin/Makefile:87: warning: ignoring old commands for target `clean'

=> Running target e2e-test Sun May 28 03:32:03 PDT 2017

../misc/scripts/build.sh e2e-dkrVolDriver-test "10.110.48.56" "10.110.48.54" "10.110.48.50"
make: Entering directory `/go/src/github.com/vmware/docker-volume-vsphere/vmdk_plugin'
Makefile:152: warning: overriding commands for target `clean'
../plugin/Makefile:87: warning: ignoring old commands for target `clean'

=> Running target e2e-dkrVolDriver-test Sun May 28 10:32:03 UTC 2017

GO15VENDOREXPERIMENT=1 go test github.com/vmware/docker-volume-vsphere/tests/vmg/ -v VmGroup
=== RUN   Test
2017/05/28 10:32:05 Creating volume [vmgroup_volume_5577006791947779410] on VM [10.110.48.54]
2017/05/28 10:32:15 Listing volumes.
2017/05/28 10:32:20 Attaching volume [vmgroup_volume_5577006791947779410] on VM [10.110.48.54]
2017/05/28 10:32:23 Creating volume [vmgroup_volume_8674665223082153551] on VM [10.110.48.54]
2017/05/28 10:32:25 Destroying volume [vmgroup_volume_5577006791947779410]
2017/05/28 10:32:32 Creating volume [vmgroup_volume_4037200794235010051] on VM [10.110.48.54]
2017/05/28 10:32:36 Creating volume [vmgroup_volume_3916589616287113937] on VM [10.110.48.54]
2017/05/28 10:32:36 Creating volume [vmgroup_volume_6334824724549167320] on VM [10.110.48.54]
2017/05/28 10:32:38 Creating volume [vmgroup_volume_3916589616287113937] on VM [10.110.48.54]
2017/05/28 10:32:42 Destroying volume [vmgroup_volume_4037200794235010051]
2017/05/28 10:32:42 Destroying volume [vmgroup_volume_3916589616287113937]
2017/05/28 10:32:49 Creating volume [vmgroup_volume_605394647632969758] on VM [10.110.48.54]
2017/05/28 10:32:50 Listing volumes.
2017/05/28 10:32:52 Inspecting volume [vmgroup_volume_605394647632969758] on VM [10.110.48.54]
2017/05/28 10:32:52 Destroying volume [vmgroup_volume_605394647632969758]
2017/05/28 10:32:53 Destroying volume [vmgroup_volume_605394647632969758]
2017/05/28 10:33:03 Creating volume [vmgroup_volume_2775422040480279449] on VM [10.110.48.54]
2017/05/28 10:33:05 Listing volumes.
2017/05/28 10:33:05 Attaching volume [vmgroup_volume_2775422040480279449] on VM [10.110.48.54]
2017/05/28 10:33:06 Sleep for 2 seconds
2017/05/28 10:33:08 Destroying volume [vmgroup_volume_2775422040480279449]
2017/05/28 10:33:15 Creating volume [vmgroup_volume_3510942875414458836] on VM [10.110.48.54]
2017/05/28 10:33:17 Attaching volume [vmgroup_volume_3510942875414458836] on VM [10.110.48.54]
2017/05/28 10:33:19 Sleep for 2 seconds
2017/05/28 10:33:21 Destroying volume [vmgroup_volume_3510942875414458836]
OK: 5 passed, 1 skipped
--- PASS: Test (86.78s)
PASS
ok      github.com/vmware/docker-volume-vsphere/tests/vmg       86.782s
make: Leaving directory `/go/src/github.com/vmware/docker-volume-vsphere/vmdk_plugin'
make[1]: Leaving directory `/home/administrator/vpl-test/docker-volume-vsphere/vmdk_plugin'

shaominchen commented on misc/scripts/build.sh:
```diff
@@ -140,6 +139,9 @@ else
     -e "VM1=$3" \
     -e "VM2=$4" \
+    -e "DS1=$DS1" \
+    -e "VM1NAME=$VM1NAME" \
```
We have already introduced govc util to retrieve VM name from IP. So these variables are not necessary.

govint replied:
I'm not sure, each test is doing something of its own for input variables. I saw the govc functions but with my ESX (6.6) keeps failing with "connection reset by peer). I'll use govc once these changes are in and I can verify with 6.5 ESX later.

shaominchen replied:
Does this issue happen on 6.0U2 and 6.5? Can you please file an issue for this? We need to make sure the test utils work on all supported platforms. @shuklanirdesh82 

Our goal is to keep the number of shared environment variables to the minimum (and use common utils to get whatever we need), because these variables must be defined when we run e2e tests locally. If there are too many variables to be defined, it will be too cumbersome to run e2e tests locally.

govint replied:
I'm using the current storage-main. I read govc is tested up to 6.5 ESX. I plan on using govc once this change is merged.

shuklanirdesh82 replied:
I would suggest to use our automation script to deploy testbed on nimbus and test out your changes locally.

shaominchen commented on tests/utils/inputparams/testparams.go:
```diff
@@ -42,10 +43,10 @@ func init() {
-func GetVolumeNameWithTimeStamp(volumeName string) string {
-	return volumeName + "_volume_" + strconv.FormatInt(time.Now().Unix(), 10)
+func GetVolumeNameWithTimeStamp(volName string) string {
+	return volName + "_volume_" + strconv.FormatInt(rand.Int63(), 10)
```
Any reason for this change? Please update comment at line 46 and remove line 49 if we decide to make this change.

govint replied:
I don't know if any one else noticed but I keep getting the same file name across multiple calls to this function and the tests break because two volume creates use the same name. The timestamp doesn't change between calls. The rand function works better. I'll make a separate function so users of the existing function can continue using that.

shaominchen replied:
We have never run into this problem. I guess this might be your environment issue. Can you try to write a small go program to verify your environment?

shaominchen commented on tests/utils/misc/misc.go:
```diff
@@ -47,3 +48,18 @@ func WaitForExpectedState(fn getVMPowerStatus, vmName, expectedState string) boo
 }
+// LogTestStart - Print a start log with given test name and
+// current time stamp
+func LogTestStart(c *C, testGroup string, testName string) {
```
I guess it might be possible to get the test group and test case name using reflection. That way it will be much easier to use these tracing functions. We can investigate this later.

govint replied:
Agree.

shaominchen commented on tests/e2e/volumepropverification_test.go:
```diff
@@ -43,6 +43,12 @@ var (
 	containerName    string
 )
 
+func TestMain(m *testing.M) {
```
Why do we need this?

govint replied:
This is the current code, I'd removed it earlier but restoring the code as is. Let the author of this code refactor this file.

shaominchen replied:
Sure.

shuklanirdesh82 commented on tests/constants/admincli/cmd.go:
```diff
@@ -46,6 +46,15 @@ const (
+	AccessSetVMgroup = vmdkopsAdmin + "vmgroup access set --name "
+
+	// AccessGetVMgroup  Modify access perms on vmgroup
+	AccessGetVMgroup = vmdkopsAdmin + "vmgroup access ls --name "
```
minor: `AccessGetVMgroup ` => `ListVmgroupAccess`

govint replied:
Modified to GetAccessForVMgroup

shuklanirdesh82 commented on tests/constants/admincli/cmd.go:
```diff
@@ -46,6 +46,15 @@ const (
 	ReplaceVMFromVMgroup = vmdkopsAdmin + "vmgroup vm replace --name="
 
+	// AccessSetVMgroup  Modify access perms on vmgroup
+	AccessSetVMgroup = vmdkopsAdmin + "vmgroup access set --name "
```
SetVmgroupAccess to keep it consistent with other const variables "Verb+Noun"

shaominchen replied:
+1

govint replied:
Fixed.

shuklanirdesh82 commented on tests/utils/dockercli/volumelifecycle.go:
```diff
@@ -30,15 +30,26 @@ import (
 func CreateVolume(ip, name string) (string, error) {
 	log.Printf("Creating volume [%s] on VM [%s]\n", name, ip)
-	return ssh.InvokeCommand(ip, dockercli.CreateVolume+"--name="+name)
+	return ssh.InvokeCommand(ip, dockercli.CreateVolume + " --name= " + name) 
```
nit: gofmt removes embedded spacing .. `dockercli.CreateVolume + " --name= " + name` looks cleaner and more readable but creates inconsistency with other formatting.

Supplying this comment to avoid future formatting change.

govint replied:
This is more readable though.

shuklanirdesh82 replied:
I agree .. for some of us `Visual studio code's gofmt plugin` auto formats hence the comment .. anyway it's nit.

shuklanirdesh82 commented on tests/utils/inputparams/testparams.go:
```diff
@@ -42,10 +43,14 @@ func init() {
+}
+
+// GetVolumeName prepares unique volume name with a random generated number
+func GetNewVolumeName(volName string) string {
```
why can not we stick with `GetVolumeNameWithTimeStamp()`?  is there any benefit for introducing this new method?

shuklanirdesh82 commented on vmdk_plugin/Makefile:
```diff
@@ -355,7 +355,7 @@ e2e-test:
 e2e-dkrVolDriver-test:
 	$(log_target)
-	$(GO) test $(E2E_Tests) -v
+	$(GO) test $(E2E_Tests) -v VmGroup
```
need to remove this file from the PR.

shuklanirdesh82 commented on tests/e2e/vmgroups_test.go:
```diff
@@ -0,0 +1,351 @@
+// After installing vmdk volume plugin/driver, volume creation should not be
+// failed very first time.
+
+// This test is going to cover the issue reported at #656
```
I guess .. this block comment needs to be revisited.

shuklanirdesh82 commented on tests/e2e/vmgroups_test.go:
```diff
@@ -0,0 +1,351 @@
+)
+
+const (
+	defaultVg                = "_DEFAULT"
```
https://github.com/vmware/docker-volume-vsphere/blob/master/tests/constants/admincli/cmd.go#L50

govint replied:
Removed.

shuklanirdesh82 commented on tests/e2e/vmgroups_test.go:
```diff
@@ -0,0 +1,351 @@
+	vg.dockerHostNames = append(vg.dockerHostNames, os.Getenv("VM2NAME"))
+	vg.host = inputparams.GetEsxIP()
+	vg.vmgroup = defVgName
+	vg.datastore1 = "_VM_DS"
```
https://github.com/vmware/docker-volume-vsphere/blob/master/tests/constants/admincli/cmd.go#L53

govint replied:
Removed.

govint commented:
Except for the make file change (left over from testing) all others are basically nits. If there isn't anything else blocking I'd prefer merging this PR rather than delay this further.

shuklanirdesh82 commented:
@govint The PR still says `SKIP CI` .. can you please remove that tag and let CI validates your testcase? I saw you posted local test run.

shuklanirdesh82 commented:
@govint Please update the summary appropriately ..

shaominchen commented on tests/e2e/vmgroups_test.go:
```diff
@@ -0,0 +1,358 @@
+func (vg *VmGroupTest) TestVmGroupVolumeCreateOnDefaultVg(c *C) {
+	misc.LogTestStart(c, vmgroupsTest, "TestVmGroupVolumeCreateOnDefaultVg")
+
+	vg.vmgroupGetVolName(c)
```
Should we do this in SetUpTest instead?

govint replied:
Yes, fixed.

shaominchen commented on tests/e2e/vmgroups_test.go:
```diff
@@ -0,0 +1,358 @@
+	c.Assert(err, IsNil, Commentf(out))
+}
+
+func (vg *VmGroupTest) vmgroupGetVolName(c *C) {
```
Nit: let's follow verb+noun naming convention if possible - simply getVolumeNames?

govint replied:
I'll do ths before the merge.

shaominchen commented on tests/utils/dockercli/volumelifecycle.go:
```diff
@@ -42,9 +42,14 @@ func CreateVolumeWithOptions(ip, name, options string) (string, error) {
+		":/vol1 --name "+containerName+ dockercli.TestContainer)
+}
+
+// InspectVolume - attach volume to container on given host
```
Nit: update this comment

govint replied:
Done

shaominchen commented on tests/utils/inputparams/testparams.go:
```diff
@@ -43,10 +51,14 @@ func init() {
 // GetVolumeNameWithTimeStamp prepares unique volume name by appending current time-stamp value
-func GetVolumeNameWithTimeStamp(volumeName string) string {
-	return volumeName + "_volume_" + strconv.FormatInt(time.Now().Unix(), 10)
+func GetVolumeNameWithTimeStamp(volName string) string {
```
We should remove this and use GetUniqueVolumeName instead in all tests. I've filed issue #1344 to track this separately.

shuklanirdesh82 replied:
@govint `return volName + "_volume_" + strconv.FormatInt(rand.Int63(), 10)` generates volume name uniquely then I am fine to address #1344 as part of this review .. it is just a simple fix and can be accommodated as part of any ongoing review. 

If agree need to remove `GetUniqueVolumeName ` and supply `strconv.FormatInt(rand.Int63(), 10)` to all 3 utils mentioned in #1344.

Please share your idea.

govint commented:
VM group tests have passed failure is in restart test and volume properties test. Pls. review and let me know any more comments. I've got to fix up some comments but all other issues and bugs identified are raised as separate issues.

pshahzeb commented on tests/utils/dockercli/volumelifecycle.go:
```diff
@@ -42,9 +42,14 @@ func CreateVolumeWithOptions(ip, name, options string) (string, error) {
-	return ssh.InvokeCommand(ip, dockercli.RunContainer+"-d -v "+volName+
-		":/vol1 --name "+containerName+
-		dockercli.TestContainer)
+	return ssh.InvokeCommand(ip, dockercli.RunContainer+" -d --rm -v "+volName+
```
adding --rm command will remove the container after the container exits.
We have test cases where we want the same container to start again after it exits.
We won't be able to simulate such test cases if we add this flag.

```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume create -d vsphere --name test2
test2
root@sc-rdops-vm02-dhcp-52-237:~# docker run  -d --rm -v test2:/vol1 busybox tail -f /dev/null
726f2d815b18ed7f514b6d1f165033302d6c1aa36121ed2a3db4513abbf8f725
root@sc-rdops-vm02-dhcp-52-237:~# docker ps -a
CONTAINER ID        IMAGE               COMMAND               CREATED             STATUS              PORTS               NAMES
726f2d815b18        busybox             "tail -f /dev/null"   5 seconds ago       Up 3 seconds                            goofy_jennings
root@sc-rdops-vm02-dhcp-52-237:~# docker stop 726f2d815b18
726f2d815b18
root@sc-rdops-vm02-dhcp-52-237:~# docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
```

At the end of the above steps, I need to start the same container again, but now its already removed.

Instead, cleaning containers could be handled in the tests itself; and anyway we are about to have a common teardown that would essentially delete containers that were created as part of a test.
I agree that adding this flag would allow us to skip specifically remove containers in cleanup in some cases. But we do have a valid cause for not having this flag.

This is also the reason the restart test failed for this PR. The teardown tried to remove the container, but it was already removed due to this flag.

govint replied:
OTOH I had the same issue in reverse, my vmgroups couldn't get removed because the container was still running and the volumes attached to the VM, so a vmgroup remove with --remove-volumes failed.

I'll remove the "rm" from here and add a function that will allow launcing and removing a container and will also make a stab at putting a doc together so we have an idea how the tests must be written.

shuklanirdesh82 replied:
How about creating a test util which attach/mounts volume with `--rm` flag to avoid adding cleanup step for such containers removal.

govint commented:
Test output,
2017/06/05 06:12:34 Finding VM name from IP Address [192.168.31.81]
2017/06/05 06:12:34 Finding VM name from IP Address [192.168.31.85]
2017/06/05 06:12:34 Finding Datastores available on ESX
2017/06/05 06:12:34 Initializing the SingleNode Config DB on esx [192.168.31.62] 
2017/06/05 06:12:36 Vmgroup vmgroup_test1 does not exist
2017/06/05 06:12:36 Creating test vmgroup vmgroup_test1
2017/06/05 06:12:37 Adding VM [photon.vmfs] to a vmgroup [vmgroup_test1] on esx [192.168.31.62] 
2017/06/05 06:12:37 Adding VM [Ubuntu.16.10] to a vmgroup [vmgroup_test1] on esx [192.168.31.62] 
2017/06/05 06:12:38 Creating test vmgroup vmgroup_test2
2017/06/05 06:12:40 stty: standard input: Inappropriate ioctl for device
Uuid            Name           Description      Default_datas..  VM_list          
--------------  -------------  ---------------  ---------------  ---------------  
11111111-111..  _DEFAULT       This is a def..  _VM_DS                            
8cc9b896-d0f..  vmgroup_test1                   datastore1       photon.vmfs,U..  
df7e7180-008..  vmgroup_test2                   vsanDatastore
2017/06/05 06:12:40 Done creating vmgroups test config.
2017/06/05 06:12:40 START:vmgroup TestVmGroupCreateAccessPrivilege Mon Jun  5 06:12:40 UTC 2017
2017/06/05 06:12:40 Creating volume [vmgroup_volume_5577006791947779410] on VM [192.168.31.81]
2017/06/05 06:12:43 Listing volumes.
2017/06/05 06:12:44 Attaching volume [vmgroup_volume_5577006791947779410] on VM [192.168.31.81]
2017/06/05 06:12:46 Removing create access for vmgroup vmgroup_test1, datastore datastore1 on esx [192.168.31.62] 
2017/06/05 06:12:47 Creating volume [vmgroup_volume_8674665223082153551] on VM [192.168.31.81]
2017/06/05 06:12:48 Failed to invoke command [docker volume  create --driver=vsphere  --name= vmgroup_volume_8674665223082153551]: exit status 1
2017/06/05 06:12:48 Enabling create access for vmgroup vmgroup_test1, datastore datastore1 on esx [192.168.31.62] 
2017/06/05 06:12:48 Destroying volume [vmgroup_volume_5577006791947779410]
2017/06/05 06:12:49 END:vmgroup TestVmGroupCreateAccessPrivilege Mon Jun  5 06:12:49 UTC 2017
2017/06/05 06:12:49 START:vmgroup TestVmGroupVerifyMaxFileSizeOnVg Mon Jun  5 06:12:49 UTC 2017
2017/06/05 06:12:49 Setting max 1gb and total 1gb for vmgroup vmgroup_test1, datastore datastore1 on esx [192.168.31.62] 
2017/06/05 06:12:50 Creating volume [vmgroup_volume_5577006791947779410] with options [-o size=1gb] on VM [192.168.31.81]
2017/06/05 06:12:54 Creating volume [vmgroup_volume_8674665223082153551] with options [-o size=1gb] on VM [192.168.31.81]
2017/06/05 06:12:54 Failed to invoke command [docker volume  create --driver=vsphere --name=vmgroup_volume_8674665223082153551 -o size=1gb]: exit status 1
2017/06/05 06:12:54 Creating volume [vmgroup_volume_6129484611666145821] with options [-o size=1023mb] on VM [192.168.31.81]
2017/06/05 06:12:54 Failed to invoke command [docker volume  create --driver=vsphere --name=vmgroup_volume_6129484611666145821 -o size=1023mb]: exit status 1
2017/06/05 06:12:54 Setting max 1gb and total 2gb for vmgroup vmgroup_test1, datastore datastore1 on esx [192.168.31.62] 
2017/06/05 06:12:55 Creating volume [vmgroup_volume_8674665223082153551] with options [-o size=1024mb] on VM [192.168.31.81]
2017/06/05 06:13:00 Destroying volume [vmgroup_volume_5577006791947779410]
2017/06/05 06:13:01 Destroying volume [vmgroup_volume_8674665223082153551]
2017/06/05 06:13:02 END:vmgroup TestVmGroupVerifyMaxFileSizeOnVg Mon Jun  5 06:13:02 UTC 2017
2017/06/05 06:13:02 START:vmgroup TestVmGroupVolumeAccessAcrossVmGroups Mon Jun  5 06:13:02 UTC 2017
2017/06/05 06:13:02 Creating volume [vmgroup_volume_5577006791947779410] on VM [192.168.31.81]
2017/06/05 06:13:06 Listing volumes.
2017/06/05 06:13:06 Removing VM [photon.vmfs] from a vmgroup [vmgroup_test1] on esx [192.168.31.62] 
2017/06/05 06:13:07 Adding VM [photon.vmfs] to a vmgroup [vmgroup_test2] on esx [192.168.31.62] 
2017/06/05 06:13:08 Inspecting volume [vmgroup_volume_5577006791947779410] on VM [192.168.31.81]
2017/06/05 06:13:09 Failed to invoke command [docker volume inspect vmgroup_volume_5577006791947779410]: exit status 1
2017/06/05 06:13:09 Destroying volume [vmgroup_volume_5577006791947779410]
2017/06/05 06:13:09 Failed to invoke command [docker volume rm vmgroup_volume_5577006791947779410]: exit status 1
2017/06/05 06:13:09 Destroying volume [vmgroup_volume_5577006791947779410]
2017/06/05 06:13:10 Removing VM [photon.vmfs] from a vmgroup [vmgroup_test2] on esx [192.168.31.62] 
2017/06/05 06:13:11 Adding VM [photon.vmfs] to a vmgroup [vmgroup_test1] on esx [192.168.31.62] 
2017/06/05 06:13:11 END:vmgroup TestVmGroupVolumeAccessAcrossVmGroups Mon Jun  5 06:13:11 UTC 2017
2017/06/05 06:13:11 START:vmgroup TestVmGroupVolumeCreate Mon Jun  5 06:13:11 UTC 2017
2017/06/05 06:13:11 Creating volume [vmgroup_volume_5577006791947779410] on VM [192.168.31.81]
2017/06/05 06:13:15 Listing volumes.
2017/06/05 06:13:16 Attaching volume [vmgroup_volume_5577006791947779410] on VM [192.168.31.81]
2017/06/05 06:13:17 Sleep for 2 seconds
2017/06/05 06:13:19 Destroying volume [vmgroup_volume_5577006791947779410]
2017/06/05 06:13:20 END:vmgroup TestVmGroupVolumeCreate Mon Jun  5 06:13:20 UTC 2017
2017/06/05 06:13:20 START:vmgroup TestVmGroupVolumeCreateOnVg Mon Jun  5 06:13:20 UTC 2017
2017/06/05 06:13:20 Removing VM [photon.vmfs] from a vmgroup [vmgroup_test1] on esx [192.168.31.62] 
2017/06/05 06:13:21 Adding VM [photon.vmfs] to a vmgroup [vmgroup_test2] on esx [192.168.31.62] 
2017/06/05 06:13:22 Creating volume [vmgroup_volume_8674665223082153551] on VM [192.168.31.81]
2017/06/05 06:13:27 Attaching volume [vmgroup_volume_8674665223082153551] on VM [192.168.31.81]
2017/06/05 06:13:30 Sleep for 2 seconds
2017/06/05 06:13:32 Destroying volume [vmgroup_volume_8674665223082153551]
2017/06/05 06:13:33 Removing VM [photon.vmfs] from a vmgroup [vmgroup_test2] on esx [192.168.31.62] 
2017/06/05 06:13:34 Adding VM [photon.vmfs] to a vmgroup [vmgroup_test1] on esx [192.168.31.62] 
2017/06/05 06:13:35 END:vmgroup TestVmGroupVolumeCreateOnVg Mon Jun  5 06:13:35 UTC 2017
2017/06/05 06:13:35 START:vmgroup TestVmGroupVolumeMobility Mon Jun  5 06:13:35 UTC 2017
2017/06/05 06:13:35 Creating volume [vmgroup_volume_5577006791947779410] on VM [192.168.31.81]
2017/06/05 06:13:38 Listing volumes.
2017/06/05 06:13:39 Attaching volume [vmgroup_volume_5577006791947779410] on VM [192.168.31.81]
2017/06/05 06:13:41 Removing VM [photon.vmfs] from a vmgroup [vmgroup_test1] on esx [192.168.31.62] 
2017/06/05 06:13:42 Removing container [vmgroupContainer_container_1496643154] on VM [192.168.31.81]
2017/06/05 06:13:42 Removing VM [photon.vmfs] from a vmgroup [vmgroup_test1] on esx [192.168.31.62] 
2017/06/05 06:13:43 Adding VM [photon.vmfs] to a vmgroup [vmgroup_test1] on esx [192.168.31.62] 
2017/06/05 06:13:44 END:vmgroup TestVmGroupVolumeAccessAcrossVmGroups Mon Jun  5 06:13:44 UTC 2017
2017/06/05 06:13:44 Removing VM [photon.vmfs] from a vmgroup [vmgroup_test1] on esx [192.168.31.62] 
2017/06/05 06:13:45 Removing VM [Ubuntu.16.10] from a vmgroup [vmgroup_test1] on esx [192.168.31.62] 
2017/06/05 06:13:46 Removing VM [photon.vmfs] from a vmgroup [vmgroup_test2] on esx [192.168.31.62] 
2017/06/05 06:13:47 Removing VM [Ubuntu.16.10] from a vmgroup [vmgroup_test2] on esx [192.168.31.62] 
2017/06/05 06:13:47 Deleting a vmgroup [vmgroup_test1] on esx [192.168.31.62]
2017/06/05 06:13:50 Deleting a vmgroup [vmgroup_test2] on esx [192.168.31.62]
2017/06/05 06:13:52 stty: standard input: Inappropriate ioctl for device
Uuid                  Name      Description           Default_datastore  VM_list  
--------------------  --------  --------------------  -----------------  -------  
11111111-1111-1111..  _DEFAULT  This is a default ..  _VM_DS
2017/06/05 06:13:52 Done cleanup of vmgroups test config.

shuklanirdesh82 commented on tests/constants/admincli/cmd.go:
```diff
@@ -52,6 +52,15 @@ const (
+	// AccessSetVMgroup  Modify access perms on vmgroup
+	SetAccessForVMgroup = vmdkopsAdmin + "vmgroup access set --name "
+
+	// AccessGetVMgroup  Modify access perms on vmgroup
```
nit: need to update the inline comment here.

shuklanirdesh82 commented on tests/utils/admincli/vmgroupmgmt.go:
```diff
@@ -54,6 +54,25 @@ func ReplaceVMFromVMgroup(ip, name, vmName string) (string, error) {
+	return ssh.InvokeCommand(ip, admincli.SetAccessForVMgroup + name + " --allow-create True --datastore " + datastore)
+}
+
+// RemoveCreateAccessForVMgroup - set allow-create access on the vmgroup
```
nit: method comment needs update

shuklanirdesh82 commented on tests/utils/dockercli/volumelifecycle.go:
```diff
@@ -171,6 +183,17 @@ func StopContainer(ip, containerName string) (string, error) {
 	return ssh.InvokeCommand(ip, dockercli.StopContainer+containerName)
 }
 
+// AttachVolume - attach volume to container on given host
```
I guess this code block refers https://github.com/vmware/docker-volume-vsphere/pull/1286#discussion_r120119292

can you please refresh the method comment as per underlying util's code?

shuklanirdesh82 commented on tests/utils/dockercli/volumelifecycle.go:
```diff
@@ -171,6 +183,17 @@ func StopContainer(ip, containerName string) (string, error) {
+	if err != nil {
+		return out, err
+	}
+	return ssh.InvokeCommand(ip, dockercli.RemoveContainer + containerName)
```
not sure L#194 and L#189 is consistent .. can you please help me to understand why do we need L194?

I would prefer test to invoke `mount volume with --rm flag`  > `detach or exit container which removes automatically`.

I would wait for your response though.

govint replied:
The --rm flag doesn't remove the container till the container exits. And the container won't exit till the command being run exits. Doing a "tail -f on /dev/null" never exits. I'm not changing that in this PR.

shuklanirdesh82 commented on tests/utils/ssh/ssh.go:
```diff
@@ -48,3 +48,4 @@ func InvokeCommandLocally(cmd string) string {
 	}
 	return strings.TrimSpace(string(out[:]))
 }
+
```
Please remove this file from the review.

shuklanirdesh82 commented on tests/utils/inputparams/testparams.go:
```diff
@@ -23,8 +23,17 @@ import (
 	"os"
 	"strconv"
 	"time"
+	"github.com/vmware/docker-volume-vsphere/tests/utils/govc"
```
CI is failing due to this import.

```
panic: runtime error: index out of range

goroutine 1 [running]:
panic(0x65a700, 0xc4200100f0)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/vmware/docker-volume-vsphere/tests/utils/ssh.init()
	/drone/src/github.com/vmware/docker-volume-vsphere/tests/utils/ssh/ssh.go:27 +0x177
github.com/vmware/docker-volume-vsphere/tests/utils/govc.init()
	/drone/src/github.com/vmware/docker-volume-vsphere/tests/utils/govc/cmd.go:68 +0x38
github.com/vmware/docker-volume-vsphere/tests/utils/inputparams.init()
	/drone/src/github.com/vmware/docker-volume-vsphere/tests/utils/inputparams/testparams.go:145 +0x47
github.com/vmware/docker-volume-vsphere/vmdk_plugin/drivers/vmdk/vmdkops_test.init()
	/drone/src/github.com/vmware/docker-volume-vsphere/vmdk_plugin/drivers/vmdk/vmdkops/cmd_test.go:50 +0x33
main.init()
	github.com/vmware/docker-volume-vsphere/vmdk_plugin/drivers/vmdk/vmdkops/_test/_testmain.go:110
```

Proposed solution:
1. move https://github.com/vmware/docker-volume-vsphere/blob/master/tests/utils/ssh/ssh.go#L27 `sshIdentity` parameters to this file 
2. make sure `os.Getenv("SSH_KEY_OPT")` not empty before splitting. 
3. have an util for generating `sshIdentity` parameters and ssh.go should call to inputParams to retrieve it. 

@govint Please let me know if you can address as part of this review otherwise I can send out review to unblock yours.

shaominchen replied:
@shuklanirdesh82 Is this a regression issue? I'd suggest to send a separate quick fix for this blocking CI issue.

shuklanirdesh82 replied:
Nope this is not a regression issue. govc library loads ssh library which initializes `sshIdentity` so above proposed fix is required to go in with this PR or with separate PR before this one. Instead of waiting for merge to master and rebase this PR .. I would like to request Govindan to supply a fix. 

@govint  Let me know what do you think? Looking forward to your response.

shaominchen commented on tests/utils/admincli/vmgroupmgmt.go:
```diff
@@ -60,7 +60,7 @@ func AddCreateAccessForVMgroup(ip, name, datastore string) (string, error) {
 }
 
-// RemoveCreateAccessForVMgroup - set allow-create access on the vmgroup
+// RemoveCreateAccessForVMgroup - remove ellow-create access on the vmgroup
```
Typo: ellow=>allow
